src/test/java/com/arpnetworking/commons/hostresolver/BackgroundCachingHostResolverTest.java

```diff
@@ -0,0 +1,120 @@
+        Assert.assertEquals("foo.example.com", backgroundCachingHostResolver.getLocalHostName());
+        Mockito.verify(_hostResolver).getLocalHostName();
+        Assert.assertEquals("foo.example.com", backgroundCachingHostResolver.getLocalHostName());
+        Mockito.verifyNoMoreInteractions(_hostResolver);
+    }
+
+    @Test
+    public void testInitialFailure() throws UnknownHostException {
+        final HostResolver backgroundCachingHostResolver = new BackgroundCachingHostResolver(
+                _hostResolver,
+                Duration.ofSeconds(10));
+        Mockito.doThrow(new UnknownHostException()).when(_hostResolver).getLocalHostName();
+        try {
+            backgroundCachingHostResolver.getLocalHostName();
+            Assert.fail("Expected exception not thrown");
+        } catch (final UnknownHostException e) {
+            // Expected exception
+        }
+    }
+
+    @Test
+    public void testBackgroundRefresh() throws UnknownHostException, InterruptedException {
+        Mockito.doReturn("foo.example.com").when(_hostResolver).getLocalHostName();
+        final HostResolver backgroundCachingHostResolver = new BackgroundCachingHostResolver(
+                _hostResolver,
+                Duration.ofSeconds(1));
+
+        // Initial call caches the result by invoking the wrapped provider
+        Assert.assertEquals("foo.example.com", backgroundCachingHostResolver.getLocalHostName());
+        Mockito.verify(_hostResolver).getLocalHostName();
+
+        // Second call within timeout does _not_ invoke the wrapped provider
+        Mockito.doReturn("foo2.example.com").when(_hostResolver).getLocalHostName();
+        Assert.assertEquals("foo.example.com", backgroundCachingHostResolver.getLocalHostName());
+        Mockito.verifyNoMoreInteractions(_hostResolver);
+
+        Thread.sleep(1500);
+
+        // Third call after the timeout invokes the wrapped provider
+        Assert.assertEquals("foo2.example.com", backgroundCachingHostResolver.getLocalHostName());
+        Mockito.verify(_hostResolver, Mockito.times(2)).getLocalHostName();
+
+        // Fourth call within the timeout does _not_ invoke the wrapped provider
+        Assert.assertEquals("foo2.example.com", backgroundCachingHostResolver.getLocalHostName());
+        Mockito.verifyNoMoreInteractions(_hostResolver);
+    }
+
+    @Test
+    public void testNoNegativeCaching() throws UnknownHostException, InterruptedException {
+        Mockito.doReturn("foo.example.com").when(_hostResolver).getLocalHostName();
+        final HostResolver backgroundCachingHostResolver = new BackgroundCachingHostResolver(
+                _hostResolver,
+                Duration.ofSeconds(1));
+
+        Assert.assertEquals("foo.example.com", backgroundCachingHostResolver.getLocalHostName());
+        Mockito.verify(_hostResolver).getLocalHostName();
+
+        Mockito.doThrow(new UnknownHostException()).when(_hostResolver).getLocalHostName();
+
+        Thread.sleep(1500);
```

Comment:
I know how that was done. However, there are no sleep statements or logical deferment in the code. The scheduled executor service does all of this. Therefore, it's not trivial to control the flow. I suppose, we could inject a custom executor service. That would do it. However, we should still have an end-to-end test at least to demonstrate the integration of the actual executor works. In principal, I completely agree with you. In practice, here, I'm not sure this is worthwhile; the complexity difference here versus the file tailer is night and day.